Use MPI for NCCL unique ID exchange by default

okkwon opened the pull request:
Now the default implementation of the channel provider uses MPI to exchange NCCL unique ID.

A few MPI functions are dynamically called just like how NCCL functions are called. With the default implementation, OpenMPI must be available in the system to run a module with collective operations. (Note that We can't use MPICH-variants because of the difference in defining MPI_Datatype and MPI_Comm.) 

The device index is overridden by the MPI rank automatically.

Here is the command line before and after.

before
```
NCCL_COMM_ID=127.0.0.1:8000 iree-run-module --cuda_nccl_default_rank=0 --cuda_nccl_default_count=2 --device=cuda://0 --module=all_gather.vmfb --function=all_gather_dim_0 & \
NCCL_COMM_ID=127.0.0.1:8000 iree-run-module --cuda_nccl_default_rank=1 --cuda_nccl_default_count=2 --device=cuda://1 --module=all_gather.vmfb --function=all_gather_dim_0
```

after
```
mpirun -np 2 iree-run-module --device=cuda --module=all_gather.vmfb --function=all_gather_dim_0
```

okkwon committed:
use MPI for NCCL unique ID exchange when available

When MPI is used, the following options are unnecessary.
- `cuda_nccl_default_rank`
- `cuda_nccl_default_count`

In addition, the device index is overriden by the MPI rank automatically.

`mpirun` can be used to run `iree-run-module`. For example,
`mpirun -np 2 iree-run-module --device=cuda ...`.

okkwon committed:
use dynamic loading for MPI

okkwon committed:
Initialize MPI in the CUDA runtime

With OpenMPI, `MPI_Init(NULL, NULL)` can be used without passing argc and argv.

cc10512 commented on build_tools/third_party/mpi/CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+  NAME
+    headers_and_libs
+  ROOT
+    "${IREE_SOURCE_DIR}/third_party/mpi/"
```
This should be derived from where the packages was found. Installing MPI using distro packages will not place the installation inside the IREE tree.

Unfortunately, FindMPI.cmake only uses `MPI_HOME` and does not export a root directory. A root can be inferred though from the include or library path.

cc10512 commented on tools/iree-run-module-main.cc:
```diff
@@ -221,6 +229,9 @@ extern "C" int main(int argc, char** argv) {
         "[ERROR] unexpected positional argument (expected none)."
         " Did you use pass a flag with a single dash ('-')?"
         " Use '--' instead.\n");
+#if IREE_USE_MPI
```
I recommend moving the `exit_code` definition up, and rather than returning early and calling `MPI_Finalize` on each branch, set the exit code, and return only once, at the end, after calling `MPI_Finalize`.

okkwon committed:
delete mpi headers

We would need to add mpi.h and its dependencieswhen the complete set of features
is needed, but for now, it is very minimal and it can be run without having it.

okkwon committed:
move the default channel provide implementation to cuda_device.c

benvanik commented on runtime/src/iree/hal/drivers/cuda/cuda_device.c:
```diff
@@ -294,6 +301,50 @@ IREE_API_EXPORT iree_status_t iree_hal_cuda_nccl_get_unique_id(
+  MPI_RETURN_IF_ERROR(syms,
+                      MPI_Comm_size(syms->ompi_mpi_comm_world, &params->count),
+                      "MPI_Comm_size");
+  ;
```
```suggestion
```

benvanik commented on runtime/src/iree/hal/drivers/cuda/cuda_device.c:
```diff
@@ -294,6 +301,50 @@ IREE_API_EXPORT iree_status_t iree_hal_cuda_nccl_get_unique_id(
+        "MPI should be loaded to use NCCL collective operations.");
+  }
+
+  // Until we have multi channel support, we only creates the default channel.
```
```suggestion
  // Until we have multi channel support, we only create the default channel.
```

benvanik commented on runtime/src/iree/hal/drivers/cuda/cuda_device.c:
```diff
@@ -294,6 +301,50 @@ IREE_API_EXPORT iree_status_t iree_hal_cuda_nccl_get_unique_id(
+IREE_API_EXPORT iree_status_t iree_hal_cuda_nccl_query_group_params(
+    void* self, iree_hal_device_t* device,
+    iree_hal_queue_affinity_t queue_affinity, iree_byte_span_t id_storage,
+    iree_hal_channel_params_t* params) {
```
```suggestion
    iree_hal_channel_params_t* out_params) {
```
(here and elsewhere)
convention to indicate that an argument is an output prefix with `out_` - this lets the caller know that it's only written and the contents are not expected to be populated prior to calling

benvanik commented on runtime/src/iree/hal/drivers/cuda/cuda_driver.c:
```diff
@@ -46,6 +46,26 @@ IREE_API_EXPORT void iree_hal_cuda_driver_options_initialize(
+}
+
+// Get the MPI world sizefrom the environmental variable. Return 0 if the
+// variable is not set.
```
```suggestion
// Gets the MPI world size from the environmental variable.
// Returns 0 if the variable is not set.
```
https://google.github.io/styleguide/cppguide.html#Function_Comments
> Function comments should be written with an implied subject of This function and should start with the verb phrase; for example, "Opens the file", rather than "Open the file"

benvanik commented on runtime/src/iree/hal/drivers/cuda/cuda_driver.c:
```diff
@@ -46,6 +46,26 @@ IREE_API_EXPORT void iree_hal_cuda_driver_options_initialize(
+
+// Get the MPI world sizefrom the environmental variable. Return 0 if the
+// variable is not set.
+iree_status_t iree_hal_cuda_get_mpi_comm_world_size_from_env(int32_t* size) {
```
```suggestion
static iree_status_t iree_hal_cuda_get_mpi_comm_world_size_from_env(int32_t* out_size) {
```
(out_* prefix as above)
(functions local to a file should be declared as static)

benvanik commented on runtime/src/iree/hal/drivers/cuda/cuda_driver.c:
```diff
@@ -70,12 +90,31 @@ static iree_status_t iree_hal_cuda_driver_create_internal(
     status = iree_hal_cuda_nccl_dynamic_symbols_initialize(host_allocator,
                                                            &driver->syms);
+    if (comm_world_size > 0) {
+      status = iree_hal_mpi_dynamic_symbols_initialize(host_allocator,
```
iree_hal_cuda_nccl_dynamic_symbols_initialize may have failed, in which case this would ignore that and overwrite the status. add a check for iree_status_is_ok &&

benvanik commented on runtime/src/iree/hal/drivers/cuda/cuda_headers.h:
```diff
@@ -7,6 +7,6 @@
+#include "cuda.h"                   // IWYU pragma: export
 #include "third_party/nccl/nccl.h"  // IWYU pragma: export
-#endif  // IREE_HAL_DRIVERS_CUDA_CUDA_HEADERS_H_
+#endif                              // IREE_HAL_DRIVERS_CUDA_CUDA_HEADERS_H_
```
add a newline so it unindents the header guard #endif

benvanik commented on runtime/src/iree/hal/drivers/cuda/dynamic_symbols.c:
```diff
@@ -163,11 +196,39 @@ iree_status_t iree_hal_cuda_nccl_dynamic_symbols_initialize(
   return status;
 }
 
+// FIXME(okkwon): it is unrelated to CUDA, but iree_hal_cuda_dynamic_symbols_t
```
yeah, we'll need to move this out to iree/hal/utils/ or something in the future and decouple it so cpu/etc can use it, but this is fine to start

benvanik reviewed (changes requested):
minor tweaks but otherwise looks great!

okkwon committed:
address PR comments

benvanik commented on runtime/src/iree/hal/drivers/cuda/api.h:
```diff
@@ -79,6 +79,13 @@ IREE_API_EXPORT void iree_hal_cuda_device_params_initialize(
+IREE_API_EXPORT iree_status_t iree_hal_cuda_nccl_query_group_params(
+    void* self, iree_hal_device_t* device,
+    iree_hal_queue_affinity_t queue_affinity, iree_byte_span_t id_storage,
+    iree_hal_channel_params_t* params);
```
```suggestion
    iree_hal_channel_params_t* out_params);
```
